File: java/03-bootstrap/src/test/java/com/fResult/bootstrap/customers/services/ApplicationContextAwareBaseClass.java
package com.fResult.bootstrap.customers.services;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;

public abstract class ApplicationContextAwareBaseClass extends BaseClass {
  private final AtomicReference<ConfigurableApplicationContext> reference = new AtomicReference<>();

  @Override
  public CustomerService getCustomerService() {
    return getCurrentContext().getBean(CustomerService.class);
  }

  protected ConfigurableApplicationContext getCurrentContext() {
    return reference.updateAndGet(
        context ->
            Optional.ofNullable(context).orElse(buildApplicationContext(getConfigurationClass())));
  }

  protected ConfigurableApplicationContext buildApplicationContext(
      Class<?> componentClass, String... profiles) {

    final var context = new AnnotationConfigApplicationContext();
    context.getEnvironment().setActiveProfiles(profiles);
    context.register(componentClass);
    context.refresh();

    return context;
  }

  protected abstract Class<?> getConfigurationClass();
}

File: kotlin/03-bootstrap/src/test/kotlin/com/fResult/bootstrap/customers/services/ApplicationContextAwareBaseClass.kt
package com.fResult.bootstrap.customers.services

import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.AnnotationConfigApplicationContext
import java.util.concurrent.atomic.AtomicReference
import kotlin.reflect.KClass

abstract class ApplicationContextAwareBaseClass : BaseClass() {
  private val reference: AtomicReference<ConfigurableApplicationContext> = AtomicReference()

  override fun getCustomerService(): CustomerService {
    return getCurrentContext().getBean(CustomerService::class.java)
  }

  private fun getCurrentContext(): ConfigurableApplicationContext {
    return reference.updateAndGet { buildApplicationContext(getConfigurationClass()) }
  }

  protected abstract fun getConfigurationClass(): KClass<*>

  private fun buildApplicationContext(
    componentClass: KClass<*>,
    vararg profiles: String,
  ): ConfigurableApplicationContext = AnnotationConfigApplicationContext().apply {
    environment.setActiveProfiles(*profiles)
    register(componentClass.java)
    refresh()
  }
}
